File: src/leetCode/java/Solution152.java
package leetCode.java;

//DP Kadane's Algorithm O(n) O(1)
class Solution152 {
    public int maxProduct(int[] nums) {
        int max = 1;
        int min = 1;
        int res = Integer.MIN_VALUE;
        for(int i=0; i<nums.length; i++){
            int temp = max;
            //不能連續故該點的最大或最小值, 可能是這個數本身或是乘上之前帶過來的
            max = Math.max(nums[i], Math.max(max * nums[i], min * nums[i]));
            min = Math.min(nums[i], Math.min(temp * nums[i], min * nums[i]));
            res = Math.max(res, max);
        }
        return res;
    }
}
/*
max = max(num, max*num, min*num)
min = min(num, max*num, min*num)
 */

File: src/leetCode/c++/Solution152.cpp
//Greedy Kadane's Algorithm O(n) O(1)
class Solution152 {
public:
    int maxProduct(vector<int>& nums) {
        int maxPro = 1, minPro = 1, res = INT_MIN;
        for(int num : nums){
            int temp = maxPro;
            maxPro = max(num, max(maxPro * num, minPro * num));
            minPro = min(num, min(temp * num, minPro * num));
            res = max(res, maxPro);
        }
        return res;
    }
};